### PR TITLE
feat: add sub page for travel-guarantee, refund taxi

### PR DESCRIPTION
**Pull request by jonasbrunvoll**
Closes https://github.com/AtB-AS/kundevendt/issues/19054

This PR includes the following:
- Page and page-module for travel-guarantee.
- On state machine to handle refund for taxi, car and other public transport modes.
- Global validation for all forms using state machine.
- Global events for all forms using state machine.

### CODE REVIEW

File: src/page-modules/contact/travel-guarantee/form-selector.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+          </ul>
+
+          <p>
+            {t(
+              PageText.Contact.travelGuarantee.layoutAgreement
+                .travelGuaranteeExceptions.exclusion,
+            )}
+          </p>
+
+          <Input
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/form-selector.tsx:66]: Should use the new `Checkbox` component to have a similar appearance.

File: src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+          event as {
+            type: 'SET_TO_STOP';
+            toStop: Line['quays'][0];
+          }
+        ).toStop,
+    }),
+    setDate: assign({
+      date: ({ event }) =>
+        (
+          event as {
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts:82]: Why do you need to use `as` here? 
Also, most of these "set" functions does the same, so should probably make an reusable function instead.

File: src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    }),
+  },
+  actors: {
+    callAPI: fromPromise(async ({ context }: any) => {
+      return await fetch('/api/travel-guarantee', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts:169]: Need to use FormData to handle attachments, but it can be updated in another PR

File: src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+        SET_IBAN: {
+          actions: 'setIBAN',
+        },
+        SET_SWIFT: {
+          actions: 'setSWIFT',
+        },
+        SET_BANK_ACCOUNT_FOREIGN: {
+          actions: 'setIsBankAccountForeign',
+        },
+      },
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts:274]: I think we could use a single "on action" here? Editing fields or something 🤔

File: src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        other: {
+          on: {
+            VALIDATE: {
+              guard: 'validateInputs',
+              target: 'readyForSubmitt',
+            },
+          },
+          tags: ['other', 'selected'],
+        },
+        readyForSubmitt: {
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts:339]:
```suggestion
        readyForSubmit: {
```

File: src/page-modules/contact/travel-guarantee/index.tsx [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+import { Checkbox } from '../components/input/checkbox';
+import ErrorMessage from '../components/input/error-message';
+
+export const RefundForm = () => {
+  const { t } = useTranslation();
+  const { lines, getLinesByMode, getQuaysByLine } = useLines();
+  const [state, send] = useMachine(fetchMachine);
+
+  // Local state to force re-render to display errors.
+  const [displayErrorsDummyState, setDisplayErrorsDummyState] = useState(true);
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/index.tsx:23]: Sounds like something is not done correctly 🤔

File: src/page-modules/contact/formInputValidator.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
+import { PageText, TranslatedString } from '@atb/translations';
```
[the remark this answers is not in the document]
Comment (jonasbrunvoll) [reply src/page-modules/contact/formInputValidator.ts:1]: 👍

File: src/page-modules/contact/travel-guarantee/form-selector.tsx (hunk 1 above)
Comment (jonasbrunvoll) [reply src/page-modules/contact/travel-guarantee/form-selector.tsx:66]: 👍

File: src/page-modules/contact/formInputValidator.ts [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+  if (!context[field]) {
+    if (!errorMessages[field]) {
+      errorMessages[field] = [];
+    }
+    errorMessages[field].push(errorMessage);
+  }
+};
+
+// Helper function to set fields in travel guarantee machine, which only uses on statemachine for all three formes.
+const setFieldsToValidateSpecialCase = (stateSubmitted: string) => {
```
Comment (mortennordseth) [inline src/page-modules/contact/formInputValidator.ts:24]: A lot of duplicate code here. Could create a "commonFields" and merge

Comment (jonasbrunvoll) [reply src/page-modules/contact/formInputValidator.ts:24]: True, will do👍

File: src/page-modules/contact/formInputValidator.ts [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+      return fields;
+  }
+};
+
+export const formInputValidator = (context: any) => {
+  const inputErrorMessages: InputErrorMessages = {};
+
+  let fieldsToValidate;
+
+  if (context.stateSubmitted) {
```
Comment (mortennordseth) [inline src/page-modules/contact/formInputValidator.ts:200]: A bit hard to understand what "stateSubmitted" is. Also, when would we have a case where "stateSubmitted" is not set and we would want to validate?

File: src/page-modules/contact/travel-guarantee/form-selector.tsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+            )}
+          </p>
+
+          <p>
+            {t(
+              PageText.Contact.travelGuarantee.layoutAgreement.ticketRefundText,
+            )}
+          </p>
+
+          <h4>
```
[the remark this answers is not in the document]
Comment (mortennordseth) [reply src/page-modules/contact/travel-guarantee/form-selector.tsx:34]: Missing typo here.

File: src/page-modules/contact/travel-guarantee/form-selector.tsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+          <Typo.p textType="body__primary">
+            {t(
+              PageText.Contact.travelGuarantee.layoutAgreement
+                .travelGuaranteeExceptions.externalFactors,
+            )}
+          </Typo.p>
+          <ul className={style.rules__list}>
+            {PageText.Contact.travelGuarantee.layoutAgreement.travelGuaranteeExceptions.examples.map(
+              (example: TranslatedString, index: number) => (
+                <li key={index}>{t(example)}</li>
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/form-selector.tsx:56]: Should this include a typography component? I think this would make the text look different from the rest.

File: src/page-modules/contact/travel-guarantee/index.tsx [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+              t(ComponentText.TransportMode.modes[val])
+            }
+            valueToId={(val: TransportModeType) => val}
+            options={['bus', 'water'] as TransportModeType[]}
+            placeholder={t(
+              PageText.Contact.inputFields.transportMode.optionLabel,
+            )}
+          />
+
+          {state.context.transportMode && (
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/index.tsx:63]: Thinking out loud here, but would it be better to have the input disabled if transport mode is not set, rather than not showing it?

File: src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+  },
+  actions: {
+    setIsIntialAgreementChecked: assign({
+      isIntialAgreementChecked: ({ context }) =>
+        !context.isIntialAgreementChecked,
+    }),
+    setIsBankAccountForeign: assign({
+      isBankAccountForeign: ({ context }) => !context.isBankAccountForeign,
+    }),
+    setStateSubmitted: assign({
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts:47]: I think you can work on the naming here. The undefined could probably be removed by proper typing.

File: src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+        phoneNumber,
+        bankAccountNumber,
+        IBAN,
+        SWIFT,
+      }: any) => {
+        return await fetch('/contact/travel-guarantee', {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json',
+          },
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/travelGuaranteeFormMachine.ts:104]:
```suggestion
```

File: src/page-modules/contact/travel-guarantee/index.tsx (hunk 12 above)
Comment (jonasbrunvoll) [reply src/page-modules/contact/travel-guarantee/index.tsx:63]: Maybe that is a good alternative. The reason for why it is hided is because transportMode must be set before being able to set Line. But i could take a quick look.

File: src/translations/pages/contact.ts [hunk 15]
```diff
@@ -324,4 +467,218 @@ export const Contact = {
...
+          name: _(
+            'Stoppa ikke på holdeplassen',
+            'Did not stop at the stop',
+            'Stoppa ikkje på haldeplassen',
+          ),
+        },
+        {
+          id: 'incorrectAppInformation',
+          name: _(
+            'Feil informasjon app eller reiseplanlegger',
```
Comment (mortennordseth) [inline src/translations/pages/contact.ts:659]:
```suggestion
            'Feil informasjon i app eller reiseplanlegger',
```

File: src/translations/pages/contact.ts [hunk 16]
```diff
@@ -324,4 +467,218 @@ export const Contact = {
...
+          name: _(
+            'Mistet neste transportmiddel',
+            'Lost next mode of transport',
+            'Mista neste transportmiddel',
+          ),
+        },
+        {
+          id: 'didNotStopAtStop',
+          name: _(
+            'Stoppa ikke på holdeplassen',
```
Comment (mortennordseth) [inline src/translations/pages/contact.ts:651]:
```suggestion
            'Stoppet ikke på holdeplassen',
```

File: src/translations/pages/contact.ts [hunk 17]
```diff
@@ -254,6 +255,148 @@ export const Contact = {
...
+        info: _(
+          'Hvis du ikke kommer frem med et annet rutetilbud, refunderer vi utgifter til drosje i henhold til gjeldende regler.',
+          'If you do not provide an alternative route offer, we will refund taxi expenses according to the applicable regulations.',
+          'Viss du ikkje kjem fram med eit anna rutetilbod, refunderer vi utlegg til drosje etter gjeldande reglar.',
+        ),
+      },
+
+      refundCar: {
+        description: _(
+          'Jeg ønsker refusojon for bruka av bil',
```
Comment (mortennordseth) [inline src/translations/pages/contact.ts:344]:
```suggestion
          'Jeg ønsker refusjon for bruk av bil',
```

File: src/translations/pages/contact.ts [hunk 18]
```diff
@@ -254,6 +255,148 @@ export const Contact = {
...
+          'Reisegarantien omfattar heller ikkje tap som følge av forseinkinga, som for eksempel mista tannlegetime, jobbavtale, togavgang eller flyavgang.',
+        ),
+        checkbox: _('Jeg forsår', 'I understand', 'Eg forstår'),
+      },
+    },
+
+    subPageTitles: {
+      refundTaxi: {
+        description: _(
+          'Jeg ønsker refusojon for drosje',
```
Comment (mortennordseth) [inline src/translations/pages/contact.ts:331]:
```suggestion
          'Jeg ønsker refusjon for drosje',
```

File: src/translations/pages/contact.ts [hunk 19]
```diff
@@ -254,6 +255,148 @@ export const Contact = {
...
+            'større arrangement eller andre trafikale forhold som i stor grad rammar kollektivtrafikken',
+          ),
+          _('pandemi', 'pandemic', 'pandemi'),
+        ],
+        exclusion: _(
+          'Reisegarantien omfatter heller ikke tap som følge av forsinkelsen, som for eksempel mistet tannlegetime, jobbavtale, togavgang eller flyavgang.',
+          'The travel guarantee does not cover losses resulting from the delay, such as missed dental appointments, job agreements, train departures, or flight departures.',
+          'Reisegarantien omfattar heller ikkje tap som følge av forseinkinga, som for eksempel mista tannlegetime, jobbavtale, togavgang eller flyavgang.',
+        ),
+        checkbox: _('Jeg forsår', 'I understand', 'Eg forstår'),
```
Comment (mortennordseth) [inline src/translations/pages/contact.ts:324]:
```suggestion
        checkbox: _('Jeg forstår', 'I understand', 'Eg forstår'),
```

File: src/page-modules/contact/travel-guarantee/index.tsx (hunk 12 above)
Comment (mortennordseth) [reply src/page-modules/contact/travel-guarantee/index.tsx:63]: Can look into it in another PR since we need to update the other forms as well

File: src/page-modules/contact/formInputValidator.ts (hunk 9 above)
Comment (jonasbrunvoll) [reply src/page-modules/contact/formInputValidator.ts:127]: Will update the variable to be more specific.

File: src/page-modules/contact/formInputValidator.ts [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+        inputErrorMessages,
+        PageText.Contact.inputFields.bankAccountNumber.errorMessages.empty,
+      );
+    }
+  }
+
+  // Populate context.errorMessages
+  context.errorMessages = inputErrorMessages;
+
+  console.log('inputErrorMessages:', inputErrorMessages);
```
Comment (mortennordseth) [inline src/page-modules/contact/formInputValidator.ts:227]:
```suggestion
```

File: src/page-modules/contact/travel-guarantee/index.tsx [hunk 21]
```diff
@@ -0,0 +1,449 @@
...
+import { Checkbox } from '../components/input/checkbox';
+import ErrorMessage from '../components/input/error-message';
+
+export const RefundForm = () => {
+  const { t } = useTranslation();
+  const { getLinesByMode, getQuaysByLine } = useLines();
+  const [state, send] = useMachine(fetchMachine);
+
+  // Local state to force re-render to display errors.
+  const [forsceRerender, setForsceRerender] = useState(false);
```
Comment (mortennordseth) [inline src/page-modules/contact/travel-guarantee/index.tsx:23]:
```suggestion
  const [forceRerender, setForceRerender] = useState(false);
```

I'll take a look at the use-case for this later, but won't block merge for now.

**Review by mortennordseth — APPROVED** [review]
